smartmoneyconcepts/SMC.py

```diff
@@ -121,34 +122,83 @@ def highs_lows(cls, ohlc: DataFrame) -> Series:
                             highs_lows[i - 1] = 1
                         else:
                             highs_lows[i + 1] = 1
                 if highs_lows[i] == -1:
                     previous_low = ohlc["low"][i - 1]
                     current_low = ohlc["low"][i]
                     next_low = ohlc["low"][i + 1]
-                    if (previous_low < current_low and highs_lows[i-1] == 0) or (next_low < current_low and highs_lows[i+1] == 0):
+                    if (previous_low < current_low and highs_lows[i - 1] == 0) or (
+                        next_low < current_low and highs_lows[i + 1] == 0
+                    ):
                         highs_lows[i] = 0
                         still_adjusting = True
-                        if (previous_low < next_low and highs_lows[i-1] == 0):
+                        if previous_low < next_low and highs_lows[i - 1] == 0:
                             highs_lows[i - 1] = -1
                         else:
                             highs_lows[i + 1] = -1
 
-        levels = np.where(highs_lows != 0, np.where(highs_lows == 1, ohlc["high"], ohlc["low"]), np.nan)
+        levels = np.where(
+            highs_lows != 0,
+            np.where(highs_lows == 1, ohlc["high"], ohlc["low"]),
+            np.nan,
+        )
 
         highs_lows = pd.Series(highs_lows, name="HighsLows")
         levels = pd.Series(levels, name="Levels")
 
         return pd.concat([highs_lows, levels], axis=1)
 
     @classmethod
-    def bos_choch(cls, ohlc: DataFrame, close_break=True, filter_liquidity=False) -> Series:
+    def swing_tops_bottoms(
+        cls, ohlc: DataFrame, swing_length: int = swing_length
+    ) -> Series:
+
+        swing_tops_bottoms = np.zeros(len(ohlc), dtype=np.int32)
+        swing_type = 0
+        prev_swing_type = 0
+
+        # Calculate the highest high and lowest low for the specified length
+        upper = ohlc["high"].rolling(window=swing_length).max()
+        lower = ohlc["low"].rolling(window=swing_length).min()
+
+        # Concatenate upper and lower to df
+        ohlc["upper"] = upper
+        ohlc["lower"] = lower
+
+        # Iterate over each index in the dataframe
+        for i in range(len(ohlc)):
+            try:
+                # Determine the swing type
+                if ohlc["high"].iloc[i] > upper.iloc[i + swing_length]:
+                    swing_type = 0
+                elif ohlc["low"].iloc[i] < lower.iloc[i + swing_length]:
+                    swing_type = 1
+
+                # Check if it's a new top or bottom
+                if swing_type == 0 and prev_swing_type != 0:
+                    swing_tops_bottoms[i] = 1
+                elif swing_type == 1 and prev_swing_type != 1:
+                    swing_tops_bottoms[i] = -1
+
+                # Update the previous swing type
+                prev_swing_type = swing_type
+            except IndexError:
+                pass
+
+        return pd.Series(swing_tops_bottoms)
```

Comment:
You need to name the series.
and you also need to add the Levels data (what pip the high/low is at)

Reply:
ok